dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h

```diff
@@ -0,0 +1,209 @@
+                       bool norm_filters = true)
+      : nfilter_(nfilter), nfft_(nfft), sample_rate_(sample_rate)
+      , freq_low_(freq_low), freq_high_(freq_high > 0 ? freq_high : sample_rate / 2)
+      , norm_filters_(norm_filters) {
+    mel_low_ = MelScale::hz_to_mel(freq_low_);
+    mel_high_ = MelScale::hz_to_mel(freq_high_);
+    hz_step_ = sample_rate_ / nfft_;
+    mel_delta_ = (mel_high_ - mel_low_) / (nfilter_ + 1);
+
+    int64_t fftbin = 0;
+    int64_t fftbin_size = nfft_ / 2 + 1;
+    weights_down_.resize(fftbin_size);
+    intervals_.resize(fftbin_size);
+    norm_factors_.resize(nfilter_, T(1));
+    T mel0 = mel_low_, mel1 = mel_low_ + mel_delta_;
+    T f = centered_freqs_ ? T(0.5) * hz_step_ : T(0);
+    int last_interval = nfilter_;
+    for (int64_t interval = 0; interval <= last_interval;
+         interval++, mel0 = mel1, mel1 += mel_delta_) {
+      if (interval == last_interval)
+        mel1 = mel_high_;
+      T f0 = MelScale::mel_to_hz(mel0), f1 = MelScale::mel_to_hz(mel1);
+      if (norm_filters_ && interval < nfilter_) {
+        // Filters are normalized so that they have constant energy per band
+        T f2 = MelScale::mel_to_hz(mel1 + mel_delta_);
+        norm_factors_[interval] = T(2) / (f2 - f0);
+      }
+      T slope = T(1) / (f1 - f0);
+      for (; fftbin < fftbin_size && f < f1; fftbin++, f += hz_step_) {
+        weights_down_[fftbin] = (f1 - f) * slope;;
+        intervals_[fftbin] = interval;
+      }
+    }
+  }
+
+  void Compute(T* out, const T* in, int64_t nwindows,
+               int64_t out_stride = -1, int64_t in_stride = -1) override {
+    if (out_stride <= 0)
+      out_stride = nwindows;
+
+    if (in_stride <= 0)
+      in_stride = nwindows;
+
+    std::memset(out, 0, sizeof(T) * nfilter_ * nwindows);
+    int fftbin_size = nfft_ / 2 + 1;
+    for (int64_t fftbin = 0; fftbin < fftbin_size; fftbin++) {
+      auto *in_row_start = in + fftbin * in_stride;
+      auto filter_up = intervals_[fftbin];
+      auto weight_up = T(1) - weights_down_[fftbin];
+      auto filter_down = filter_up - 1;
+      auto weight_down = weights_down_[fftbin];
+
+      if (filter_down >= 0) {
+        if (norm_filters_)
+          weight_down *= norm_factors_[filter_down];
+        auto *out_row_start = out + filter_down * out_stride;
+        for (int t = 0; t < nwindows; t++) {
+          out_row_start[t] += weight_down * in_row_start[t];
+        }
+      }
+
+      if (filter_up < nfilter_) {
+        if (norm_filters_)
+          weight_up *= norm_factors_[filter_up];
+        auto *out_row_start = out + filter_up * out_stride;
+        for (int t = 0; t < nwindows; t++) {
+          out_row_start[t] += weight_up * in_row_start[t];
+        }
+      }
+    }
+  }
+
+ private:
+  int nfilter_;
+  int nfft_;
+  T sample_rate_;
+  T freq_low_;
+  T freq_high_;
+  bool norm_filters_;
+
+  T mel_low_;
+  T mel_high_;
+  T hz_step_;
+  T mel_delta_;
```

Comment:
Many of these fields are not used beyond the scope of the constructor.

Reply:
done